### DISCUSSION TITLE
Driving or advanced _driving?

### DISCUSSION THREAD

**Discussion by kmamor** [gama-platform/gama#218] [Q&A]
I've become quite confused as to what is the current skill available for driving models. If I search the GAMA platform for advanced driving (referenced in certain packaged models) I go here ([https://gama-platform.org/wiki/1.8.1/UsingDrivingSkill] ) but this states clearly that this is version 1.8.1. This references `skill_road`  and `skill_road_node`. Following the link to the latest version 1.9.3 the skill is now called simply 'driving skill' and we now get `road_skill` and `intersection_skill` . 

On this page there is a link for 'application example' but it is a dead link.  At the bottom of the page there is a link 'More details about the driving skill can be found [here](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0281658)' but this takes one through to  a built using v1.8.1 . 

Can someone please direct me to a driving model built in v1.9.n? Also to what extent can I simply take the v1.8.1 model and just change key words (for example exchange 'skill_road_node' with 'intersection_skill'? 

K

**Comment by lesquoyb** [comment]
Hello,
thanks for pointing out the broken link :)
I think those name changes are mostly normalization with the rest of gama's names and you can probably just switch for the new ones. The changes brought in recent versions of the skill are mostly bug fixes so functionalities must be approximately the same.
You can find working examples of this skill in the models packaged with gama, under `Plugin models` -> `Driving Skill`:
![image](https://github.com/gama-platform/gama/assets/6374469/e77fe0a2-a268-4df5-b6e8-2a8f4a3f2839)

**Comment by kmamor** [comment]
May I post a follow-up query. I have tried running the drive_random and traffic models from the advanced models folder, using my own shapefiles. However I get the error message "cannot evaluate location as the target agent is nil". This refers to a vehicle and arises on calling the line   `location <- one_of(non_deadend_nodes).location;` 

This suggests that my node file does not contain any road_outs  [`non_deadend_nodes <- intersection where !empty(each.roads_out);`] 

Before creating the roads I used 

```
 list<geometry> var0 <- clean_network(shp_roads.contents, 100, true, true);
		create road from: var0 with:[type::string(get("highway")),num_lanes::int(read("lanes")), speed::string(read("maxspeed"))] {
```

Which I thought would ensure a complete graph. Is the problem the way I'm forming the node file? 

I am using OSM via QGIS and create it by taking the highways layer / dissolve / multi part to single part/ remove duplicate geometries / line intersections.   I then combine that new layer with a layer of highways type = traffic_signals. 

Any advice welcome.

**Reply by lesquoyb** [reply]
Can't be of much help on this one but have you tried playing with the [clean_network](https://gama-platform.org/wiki/next/OperatorsBC#clean_network) operator ? It can often help when importing real data by recreating nodes and connections that do not exist explicitly if I remember correctly

**Reply by kmamor** [reply]
If I look at the node.shp file used in the drive_random model (packaged with GAMA) it comes with only two attributes: "type" and "crossing". I have tried to use a node file    it has attributes such as : "osm_type" (all 'node') "highway" (either 'traffic_signal' or 'crossing'). Just like the packaged file it does not contain any information associating it with the roads crossing it.

How does GAMA populate the 'roads_in' and 'roads_out' variables of  the intersections with 'intersection_skill'. I see in drive_random the intersection agents have these variables populated when initialised but that does not seem to be working for me. 

Relevant code extracts: 

```
global {
	
file shp_roads <- shape_file("../includes/roads3.shp");
file shp_intersections <- shape_file("../includes/intersect2.shp");
geometry shape <- envelope(shp_roads) + 50;
graph road_network;
init {
	list<geometry> var0 <- clean_network(shp_roads.contents, 100, true, true);

		create road from: var0 with:[type::string(get("highway")),num_lanes::int(read("lanes")), speed::string(read("maxspeed"))] {
				//create road in opposite direction
				create road {
				num_lanes <- myself.num_lanes;
				shape <- polyline(reverse(myself.shape.points));
				maxspeed <- myself.maxspeed;
				speed <- myself.speed;
				linked_road <- myself;
				myself.linked_road <- self;
						}
					}
	create intersection from: shp_intersections with:[is_traffic_signal::(read("highway") = "traffic_signals")]{
					}
	road_network <- as_driving_graph(road,intersection);
		}
}

species intersection skills:[intersection_skill]{
	rgb color <- #black;
	bool is_traffic_signal;
		aspect base {
			draw circle(50) color: color;
		}
	}
	
species road skills:[road_skill]{
	string type;
	string speed;
	rgb color <- #black;
	aspect base {
		draw shape color: color;
		}
}

experiment MK_traffic type:gui {
	
	output  {
		display map{
			species road aspect: base;
			species intersection aspect: base;
						} 
					}
	}
```
